lib/CPathfinder.cpp

```diff
@@ -451,53 +645,40 @@ void CGPath::convert( ui8 mode )
 	}
 }
 
-CPathsInfo::CPathsInfo( const int3 &Sizes )
-:sizes(Sizes)
+CPathsInfo::CPathsInfo(const int3 &Sizes)
+	: sizes(Sizes)
 {
 	hero = nullptr;
-	nodes = new CGPathNode**[sizes.x];
+	nodes.resize(boost::extents[sizes.x][sizes.y][sizes.z][ELayer::NUM_LAYERS]);
```

Comment:
This does not really removes memory fragmentation. You still allocate (x*y*z*num_layers + 1) blocks of memory potentially causing fragmentation & cache misses.

What you need to do is store in multimap not pointers to nodes but nodes themselves. This will result in huge (several Mb's) singular block of memory. This may seems to be too much but for a game that loads 100+ Mb of images this is not that bad.

And while pontentially not all nodes are actually used performing some sort of dynamic allocation to allocate only necessary nodes will cause CPU performance - more checks for node presence, less cache-friendly, costly calls to "new", etc.